### PR TITLE
[5.1] Added new array functions

### PR DESCRIPTION
**Pull request by yeexel**
**Arr::getIndexedByKeys**

_Test data_

```
$data = [
    ['rating' => 5, 'stars' => 5, 'language' => 'php'],
    ['rating' => 5, 'stars' => 6, 'language' => 'js'],
    ['rating' => 4, 'stars' => 4, 'language' => 'css'],
];
```

_Result_

```
Arr::getIndexedByKeys($data, ['rating', 'stars'])

[
    5 => [
        5 => ['rating' => 5, 'stars' => 5, 'language' => 'php'],
        6 => ['rating' => 5, 'stars' => 6, 'language' => 'js'],
    ],
    4 => [
        4 => ['rating' => 4, 'stars' => 4, 'language' => 'css'],
    ],
]
```

**Arr::getIndexedValues**

_Test data_

```
$data = [
    ['rating' => 5, 'stars' => 5, 'language' => 'php'],
    ['rating' => 5, 'stars' => 6, 'language' => 'js'],
    ['rating' => 4, 'stars' => 4, 'language' => 'css'],
];
```

_Result_

```
Arr::getIndexedValues($data, 'language', 'stars')

['php' => 5, 'js' => 6, 'css' => 4]

Arr::getIndexedValues($data, 'language', ['stars', 'rating'])

[
    'php' => ['stars' => 5, 'rating' => 5],
    'js'    => ['stars' => 6, 'rating' => 5],
    'css' => ['stars' => 4, 'rating' => 4],
]
```

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
https://styleci.io/analyses/qJevaX

File: src/Illuminate/Support/Arr.php [hunk 1]
```diff
@@ -456,4 +456,68 @@ public static function where($array, callable $callback)
 
         return $filtered;
     }
+
+    /**
+     * Index the array by array of keys
+     *
+     * @param array $data
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Arr.php]: please review every phpdoc since they're all incorrectly formatted

**Comment by yeexel** [comment]
Fixed code style.

File: src/Illuminate/Support/Arr.php [hunk 2]
```diff
@@ -456,4 +456,69 @@ public static function where($array, callable $callback)
 
         return $filtered;
     }
+
+    /**
+     * Index the array by array of keys.
+     *
+     * @param array $data
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Arr.php]: all phpdoc formatting still wrong

Comment (yeexel) [inline src/Illuminate/Support/Arr.php]: What's wrong with it?)

Comment (rkgrep) [inline src/Illuminate/Support/Arr.php]: http://laravel.com/docs/5.1/contributions#coding-style

Comment (yeexel) [inline src/Illuminate/Support/Arr.php]: Thanks @rkgrep

File: src/Illuminate/Support/Arr.php [hunk 3]
```diff
@@ -456,4 +456,69 @@ public static function where($array, callable $callback)
...
+     * @param  array  $array
+     * @param  string  $keyName
+     * @param  string|array  $valueName
+     * @return array
+     */
+    public static function getIndexedValues(array $array, $keyName, $valueName)
+    {
+        $result = [];
+        foreach ($array as $value) {
+            if (is_array($valueName)) {
```
Comment (rkgrep) [inline src/Illuminate/Support/Arr.php]: Loop-invariant code - you should better move this condition out of `foreach` and make two different cycles

File: src/Illuminate/Support/Arr.php [hunk 4]
```diff
@@ -456,4 +456,69 @@ public static function where($array, callable $callback)
...
+        return $result;
+    }
+
+    /**
+     * @param  array  $result
+     * @param  array  $toSave
+     * @param  array  $keys
+     * @param  bool|true $unique
+     */
+    protected static function indexByKeys(array &$result, array $toSave, array $keys, $unique = true)
```
Comment (rkgrep) [inline src/Illuminate/Support/Arr.php]: I don't see any sense in separating this method.

File: tests/Support/SupportArrTest.php [hunk 5]
```diff
@@ -248,4 +271,119 @@ public function testForget()
...
+    public function testGetIndexedValuesString()
+    {
+        $this->assertEquals(
+            [
+                'php'   => 5,
+                'js'    => 6,
+                'css'   => 4,
+                'scss'  => 4,
+            ],
+            Arr::getIndexedValues($this->testData, 'language', 'stars')
```
Comment (rkgrep) [inline tests/Support/SupportArrTest.php]: This is same as `Arr::pluck($this->testData, 'stars', 'language')`, isn't it?

**Comment by yeexel** [comment]
@rkgrep thanks

The only thing you would need to do is to add array_flip

    array_flip(Arr::pluck($this->testData, 'stars', 'language'))

Then it will function the same as getIndexedValues method

**Comment by rkgrep** [comment]
`Arr::pluck` doesn't need `array_flip` usage since you can change argument places
````php
Arr::pluck($array, 'a', 'b') == array_flip(Arr::pluck($array, 'b', 'a'));
````

**Comment by yeexel** [comment]
Got it.

**Comment by lucasmichot** [comment]
Would make sense to move this function to `\Illuminate\Support\Collection`.

File: src/Illuminate/Support/Arr.php (hunk 4 above)
Comment (lucasmichot) [inline src/Illuminate/Support/Arr.php]: I agree - it makes no sense

**Comment by JosephSilber** [comment]
I personally think this whole thing is overkill, and shouldn't be in a general-purpose library.

Just my two cents.

File: src/Illuminate/Support/Arr.php [hunk 6]
```diff
@@ -456,4 +456,45 @@ public static function where($array, callable $callback)
...
+        $result = [];
+
+        foreach ($data as $value) {
+            static::indexByKeys($result, $value, $keys, $unique);
+        }
+
+        return $result;
+    }
+
+    /**
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Arr.php]: missing short description

File: src/Illuminate/Support/Arr.php [hunk 7]
```diff
@@ -456,4 +456,45 @@ public static function where($array, callable $callback)
...
+     * @param  array  $result
+     * @param  array  $toSave
+     * @param  array  $keys
+     * @param  bool|true $unique
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Arr.php]: bool|true is wrong because true is a subset of bool

Comment (GrahamCampbell) [inline src/Illuminate/Support/Arr.php]: just put `bool`

File: src/Illuminate/Support/Arr.php [hunk 8]
```diff
@@ -456,4 +456,45 @@ public static function where($array, callable $callback)
 
         return $filtered;
     }
+
+    /**
+     * Index the array by array of keys.
+     *
+     * @param  array  $data
+     * @param  array  $keys
+     * @param  bool|true  $unique
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Arr.php]: same

File: src/Illuminate/Support/Arr.php [hunk 9]
```diff
@@ -456,4 +456,45 @@ public static function where($array, callable $callback)
...
+
+        return $result;
+    }
+
+    /**
+     * @param  array  $result
+     * @param  array  $toSave
+     * @param  array  $keys
+     * @param  bool|true $unique
+     */
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Arr.php]: missing `@return` annotation

**Comment by taylorotwell** [comment]
We typically only add stuff we use in the core.

**Pull request closed without merging**